### ISSUE TITLE
TimeService rapidly falls behind actual time.

### ISSUE THREAD

**Issue by greg-barnett** [aio-libs/aiohttp#2176]
TimeService falls behind actual time by about half a second every second.  It is supposed to reset every 10 minutes, but doesn't.

More details and a fix are available at:
https://github.com/greg-barnett/aiohttp/commit/de0cc62f1b79d9920461e54610350dea548103c7

**Comment by asvetlov** [comment]
Hmm. 
On what load you've got 50% accuracy?
I suspect it's possible only on overwhelmed event loop (or code with long blocking synchronous calls).
Also what problem it produces?
Time service is used for measuring relative slow timeouts for things like automatic socket disconnection etc. If socket will be disconnected a little later -- no problem at all!

**Comment by greg-barnett** [comment]
TimeService falling behind is not load dependent.

loop_time() (which is what I'm guessing is used for timeouts) will never fall behind by more than about 1 second.

time() and strtime() will fall behind, no matter what the load is.  strtime() falling behind causes Date headers in responses to be incorrect (this is how I discovered this).

See https://gist.github.com/greg-barnett/00b67d3b098bc782db642b068ad4b0d3 for a test script, and output from the master branch and my bugfix.

**Comment by socketpair** [comment]
Sorry, it's not documented why we have TimeService either. Why not to use just `time.time()` ?

**Comment by asvetlov** [comment]
It is implementation detail, not part of public API.
The main idea is: formatting time for `Date` HTTP header is expensive operation, it might take longer than preparing and sending response.

But the header has 1 second accuracy. We could amortize formatting time by caching a string for 1 second.

**Comment by asvetlov** [comment]
Sorry, I did mislead: time service is never used for timeout calculations but for datatime formatting only.

**Comment by socketpair** [comment]
So, caching may be implemented MUCH easier. Why not ?

```python
_cached_time = None
_cached_time_formatted = None
def format_current_datetime():
    global _cached_time
    global _cached_time_formatted
    now = int(time.time())
    if now != _cached_time:
        _cached_time_formatted = some_format(now)
    return _cached_time_formatted
```

**Comment by fafhrd91** [comment]
+1

**Comment by greg-barnett** [comment]
TimeService is also used in _process_keepalive in RequestHandler, but that should be replaceable with a call to loop.time()

**Comment by greg-barnett** [comment]
This commit has a version of TimeService that only caches the formatted string for use in Date headers.
https://github.com/greg-barnett/aiohttp/commit/eef354e79d5825f2ff63e2cdc34fb6b3b87e3975

**Comment by fafhrd91** [comment]
if we make decision to simplify TimeService, then it should be removed at all.

**Comment by greg-barnett** [comment]
This commit has slightly cleaner/faster formatting
https://github.com/greg-barnett/aiohttp/commit/51e793dcf71df5d6cba5050170f6b056f7a468e9

I agree that getting rid of TimeService completely probably makes sense, but I don't know the codebase well enough to know where the functionality should reside.  I don't want to add global variables if it can be avoided.

**Comment by greg-barnett** [comment]
Doing some more benchmarking, I'm not even convinced that caching is required.


import time
import timeit
timeit.timeit("time.strftime('%a, %d %b %Y %H:%M:%S GMT', time.gmtime())", number=10000)
0.016672083060257137

**Comment by fafhrd91** [comment]
it is, 5-7% cpu

**Comment by greg-barnett** [comment]
@fafhrd91 do you have any benchmarking code that shows TimeService's caching of formatted date makes any significant performance difference?  I have not been able to replicate any conditions where it has a noticeable impact.

**Comment by fafhrd91** [comment]
I don't have benchmarks now, but I did a lot of benchmarking when worked on performance.

**Comment by greg-barnett** [comment]
https://github.com/aio-libs/aiohttp/pull/2182
Meant to associate pull request with this issue.

**Comment by asvetlov** [comment]
I hope the problem is fixed in 2.3
@greg-barnett @socketpair can we close the issue?

**Comment by greg-barnett** [comment]
I'm currently on medical leave and can't devote any time to verifying the fix, so I will defer to @socketpair regarding closing this issue.
